server/src/main/java/org/opensearch/plugins/MapperServiceProvider.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugins;
+
+import org.opensearch.common.annotation.PublicApi;
+import org.opensearch.core.index.Index;
+import org.opensearch.index.mapper.MapperService;
+
+/**
+ * A functional interface for retrieving {@link MapperService} instances.
+ * This interface is intended for use by plugins that require access
+ * to field mappings for specific indices.
+ *
+ * @opensearch.internal
+ */
+@FunctionalInterface
+@PublicApi(since = "2.18.0")
+public interface MapperServiceProvider {
```

Comment:
I am not sure if this is the best way to do it, that too exposing this inside a plugin package. It basically tells anyone to implement a MapperServiceProvider implementation, but underneath it returns a concrete `MapperService` class
If the only intention is to expose this MapperService somehow, one way would be to pass it inside MapperAwarePlugin via param `Function<Index, MapperService> `.

Reply:
Thanks @jainankitk  I think we should not pursue this route, see please https://github.com/opensearch-project/OpenSearch/pull/16110#issuecomment-2385769023

Reply:
+1 Exposing this interface seems un-necessary to me as well to expose the MapperService.

@jainankitk @dzane17 : Please take a look [MappingStats.java](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/admin/cluster/stats/MappingStats.java#L71-L95) for reference which also iterates over all IndexMappings to get fieldType. It doesn't use MapperService and the benefit is that it can be executed anywhere even when shards of an index are not assigned on that node including Dedicated Coordinator node. Though this is very lightweight logic to get fieldType, so not sure if it serves your use case properly. You can choose to enhance ```MappingVisitor``` class if needed. QueryInsightPlugin can maintain a cache of fields and refresh it when an index is create/deleted/updated.

```
        Map<String, IndexFeatureStats> fieldTypes = new HashMap<>();
        for (IndexMetadata indexMetadata : state.metadata()) {
            Set<String> indexFieldTypes = new HashSet<>();
            MappingMetadata mappingMetadata = indexMetadata.mapping();
            if (mappingMetadata != null) {
                MappingVisitor.visitMapping(mappingMetadata.getSourceAsMap(), fieldMapping -> {
                    String type = null;
                    Object typeO = fieldMapping.get("type");
                    if (typeO != null) {
                        type = typeO.toString();
                    } else if (fieldMapping.containsKey("properties")) {
                        type = "object";
                    }


                    if (type != null) {
                        IndexFeatureStats stats = fieldTypes.computeIfAbsent(type, IndexFeatureStats::new);
                        stats.count++;
                        if (indexFieldTypes.add(type)) {
                            stats.indexCount++;
                        }
                    }
                });
            }
        }
        return new MappingStats(fieldTypes.values());
```

Reply:
Thanks @shwetathareja! We will include logic based on this approach in the `QueryInsightsPlugin`. @dzane17 - Can you make this PR limited to `successfulSearchShardIndices` changes?